### PR TITLE
feat(core): diesel models and db interface changes for authentication table

### PR DESCRIPTION
**Pull request by hrithikesh026**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Created authentication table and implemented interface for DB operations.

Hyperswitch currently uses the "Processor driven approach" for 3DS and is now looking to move towards "Third Party 3DS provider approach" - Using a specialized 3DS provider for checking card enrollment and 3DS authentication
Eg: Cardinal, Netcetra

The 3DS authentication can be done once and the authentication response that we get can be sent to any processor to authorize the payment.

Schema:
Screenshot 2024-02-28 at 2 49 33 PM

Schema changes: https://docs.google.com/spreadsheets/d/104KSSaByQu_2DH-n8qNTEJnaGw7FXtch8PTzX8fcMLk/edit#gid=1683151297

### Additional Changes

- [ ] This PR modifies the API contract
- [x] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## How did you test it?

This PR is only meant for creating new authentication table and doesn't interfere with any existing flows.

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/diesel_models/src/authentication.rs [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+    #[serde(with = "common_utils::custom_serde::iso8601")]
+    pub modified_at: time::PrimitiveDateTime,
+    pub error_message: Option<String>,
+    pub error_code: Option<String>,
+    pub connector_metadata: Option<serde_json::Value>,
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Queryable, Serialize, Deserialize, Insertable)]
+#[diesel(table_name = authentication)]
+pub struct AuthenticationNew {
```
Comment (SanchithHegde) [inline crates/diesel_models/src/authentication.rs:30]: You don't need `Queryable` here.

File: crates/diesel_models/src/authentication.rs [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+        error_message: Option<String>,
+        error_code: Option<String>,
+        authentication_status: common_enums::AuthenticationStatus,
+        connector_authentication_id: Option<String>,
+    },
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, AsChangeset, Queryable, Serialize, Deserialize)]
+#[diesel(table_name = authentication)]
+pub struct AuthenticationUpdateInternal {
```
Comment (SanchithHegde) [inline crates/diesel_models/src/authentication.rs:69]: You don't need `Queryable` here.

File: crates/diesel_models/src/query/authentication.rs [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+impl AuthenticationNew {
+    #[instrument(skip(conn))]
+    pub async fn insert(self, conn: &PgPooledConn) -> StorageResult<Authentication> {
+        generics::generic_insert(conn, self).await
+    }
+}
+
+impl Authentication {
+    #[instrument(skip(conn))]
```
Comment (SanchithHegde) [inline crates/diesel_models/src/query/authentication.rs:22]: Please remove `instrument` from methods in the `diesel_models::query` module and add it to the DB interface methods in the `router::db` module instead.

For more information, check: #3838

File: crates/router/src/db/authentication.rs [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+        &self,
+        authentication: storage::AuthenticationNew,
+    ) -> CustomResult<storage::Authentication, errors::StorageError> {
+        let mut authentications = self.authentications.lock().await;
+        if authentications.iter().any(|authentication_inner| {
+            authentication_inner.authentication_id == authentication.authentication_id
+        }) {
+            Err(errors::StorageError::DuplicateValue {
+                entity: "authentication_id",
+                key: None,
```
Comment (SanchithHegde) [inline crates/router/src/db/authentication.rs:91]:
```suggestion
                entity: "authentication",
                key: Some(authentication.authentication_id),
```

I believe these would be more suitable.

**Review by SanchithHegde — APPROVED** [review]

**Review by sahkal — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**